frontend/src/router.ts

```diff
@@ -170,16 +180,46 @@ export const router = new Router({
   ]
 });
 
-router.beforeEach((to, from, next) => {
-  let currentUser = firebase.auth().currentUser;
-  let requireAuth = to.matched.some(record => record.meta.requireAuth);
-  let guestOnly = to.matched.some(record => record.meta.guestOnly);
+export const routerBeforeEach: NavigationGuard = async (to, from, next) => {
+  // reverse routes so nested routes can take control
+  const requireAuthRoute = [...to.matched]
+    .reverse()
+    .find(({ meta }) => meta.auth);
+
+  if (!requireAuthRoute) return next();
+
+  const currentUser = firebase.auth().currentUser;
+  const {
+    meta: {
+      auth: { requireAuth, guestOnly, requireClaims }
+    }
+  } = requireAuthRoute;
 
   if (requireAuth && !currentUser)
     next({
       path: "/login",
       query: { redirect: to.fullPath }
     });
   else if (guestOnly && currentUser) next("/");
```

Comment:
Let's `return` after calling `next` in each case. This way we'll make the logic easier to comprehend and avoid nested `if`, `else` and `else if`.